### PR TITLE
Feature/cdssm

### PR DESCRIPTION
**Pull request by bwanglzu**
Add CDSSM model.

For #168

### CODE REVIEW

**Comment by bwanglzu** [comment]
@yangliuy the parameters can be checked through:
1. `params` property: for example, `dssm_model.params`.
2. documentation.

At the moment, all default hyper parameters = "paper". We can try our best to maintain a `consistent` set of hyper-parameters. Here the `consistent` denotes for:
1. Internal consistency: Hyper-parameters in MatchZoo models should be consistent.
2. External consistency: Hyper-parameters in MatchZoo models should be consistent with Keras and  community common naming conventions.

Besides, do you think the DSSM model implementation is correct?

**Comment by logicmd** [comment]
Hi @bwanglzu , just curious, is there any issues of CDSSM model in master branch? Will CDSSM merge back to master in the future or kept as a seperate branch?
Thanks!

File: matchzoo/models/cdssm_model.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+from matchzoo import engine
+
+from keras import Model
+from keras.layers import Input, Conv1D, MaxPooling1D, Dot, Dense, Flatten
+
+
+class CDSSMModel(engine.BaseModel):
+    """
+    Convolutional deep structured semantic model.
```
Comment (pl8787) [inline matchzoo/models/cdssm_model.py]: We should add CDSSM paper name in the comment.
Learning Semantic Representations Using Convolutional Neural Networks for Web Search
A Latent Semantic Model with Convolutional-Pooling Structure for Information Retrieval

File: matchzoo/models/cdssm_model.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        >>> model.guess_and_fill_missing_params()
+        >>> model.build()
+
+    """
+
+    @classmethod
+    def get_default_params(cls) -> engine.ParamTable:
+        """:return: model default parameters."""
+        params = super().get_default_params()
+        params['optimizer'] = 'sgd'
```
Comment (pl8787) [inline matchzoo/models/cdssm_model.py]: the default 'sgd' optimizer could not yield a good result if we do not adjust learning rate. 
'sgd' -> 'adam' is better for default parameter.

File: matchzoo/models/cdssm_model.py (hunk 1 above)
Comment (bwanglzu) [reply matchzoo/models/cdssm_model.py]: okay

File: matchzoo/models/cdssm_model.py (hunk 2 above)
Comment (bwanglzu) [reply matchzoo/models/cdssm_model.py]: let's keep the default parameters the same as original paper, and leave the hyper-parameters tunning to the user. @pl8787

**Comment by bwanglzu** [comment]
@logicmd It's for 2.0 branch only, no worries.

File: matchzoo/models/cdssm_model.py (hunk 2 above)
Comment (faneshion) [reply matchzoo/models/cdssm_model.py]: I think the default parameters is better. The better settings of parameter can be shown in the documentation of each model.

File: matchzoo/models/cdssm_model.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                   activation=self._params['activation_hidden'],
+                   kernel_initializer=self._params['w_initializer'],
+                   bias_initializer=self._params['b_initializer'])(x_in)
+        # Apply max pooling by take max at each dimension across
+        # all word_trigram features.
+        x = MaxPooling1D(pool_size=input_shape[1],
+                         padding=self._params['padding'])(x)
+        # Apply a none-linear transformation use a tanh layer.
+        x = Flatten()(x)
+        x_out = Dense(self._params['dim_fan_out'],
```
Comment (faneshion) [inline matchzoo/models/cdssm_model.py]: The number of full-connected layers should be able to customized.  At the same time, the number of hidden nodes also need to be defined by users.
